core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,137 @@
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink
+) : NSOutputStream(toMemory = Unit), NSStreamDelegateProtocol {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+            postEvent(NSStreamEventErrorOccurred)
+            sink.close()
+        }
+
+    override fun streamStatus() = if (isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            postEvent(NSStreamEventHasSpaceAvailable)
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError) return
+        status = NSStreamStatusClosed
+        sink.close()
+        runLoop = null
+        runLoopModes = listOf()
+    }
+
+    @OptIn(DelicateIoApi::class)
+    override fun write(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        try {
+            if (isClosed()) throw IOException("Underlying sink is closed.")
+            if (status != NSStreamStatusOpen) return -1
+            if (buffer == null) return -1
+
+            status = NSStreamStatusWriting
+            sink.writeToInternalBuffer {
+                it.write(buffer, maxLength.toInt())
+            }
+            status = NSStreamStatusOpen
+            return maxLength.convert()
+        } catch (e: Exception) {
+            error = e.toNSError()
+            return -1
+        }
+    }
+
+    override fun hasSpaceAvailable() = !isClosed()
+
+    @OptIn(InternalIoApi::class)
+    override fun propertyForKey(key: NSStreamPropertyKey): Any? = when (key) {
+        NSStreamDataWrittenToMemoryStreamKey -> sink.buffer.snapshotAsNSData()
+        else -> null
+    }
+
+    override fun setProperty(property: Any?, forKey: NSStreamPropertyKey) = false
+
+    private var _delegate = WeakReference<NSStreamDelegateProtocol>(this)
+    private var runLoop: NSRunLoop? = null
+    private var runLoopModes = listOf<NSRunLoopMode>()
+
+    private fun postEvent(event: NSStreamEvent) {
+        val runLoop = runLoop ?: return
+        runLoop.performInModes(runLoopModes) {
+            if (runLoop == this.runLoop) {
+                delegate?.stream(this, event)
+            }
+        }
+    }
+
+    override fun delegate() = _delegate.value
```

Comment:
Yeah, the docs are really not clear on this. And despite its [explicit requirement](https://developer.apple.com/documentation/foundation/nsstream/1418423-delegate#discussion) to make the stream its own delegate by default, this doesn't actually make sense with our subclass since our self delegate implementation does nothing, and our class isn't `open` for subclassing, in fact it's not even public. But that's the only reason I've implemented `NSStreamDelegateProtocol`, since it says to, and maybe there's some messaging magic that may be possible while using the Kotlin object within Objective-C code.

I checked the default `NSOutputStream.delegate` returns `null` by default. So I don't think the expectation is to return `self` when it's `nil`.